File: 2678-design-graph-with-shortest-path-calculator/design-graph-with-shortest-path-calculator.java
class Graph {

    int[][] graph;
    int n;

    public Graph(int n, int[][] edges) {
        graph = new int[n][n];
        this.n = n;
        for(int[] g : graph) Arrays.fill(g, (int)1e8);
        for(int i=0; i<edges.length; i++) {
            graph[edges[i][0]][edges[i][1]] = edges[i][2];
        }

        for(int i=0; i<n; i++) graph[i][i] = 0;

        for(int k=0; k<n; k++) {
            for(int i=0; i<n; i++) {
                for(int j=0; j<n; j++) {
                    if(graph[i][k] + graph[k][j] < graph[i][j]) {
                        graph[i][j] = graph[i][k] + graph[k][j];
                    }
                }
            }
        }
        
    }
    
    public void addEdge(int[] edge) {
        int e1 = edge[0];
        int e2 = edge[1];
        int cost = edge[2];

        if(graph[e1][e2] > cost) {
            //graph[e1][e2] = cost;
            updDist(e1, e2, cost);
        }
    }

    void updDist(int u, int v, int c) {
        for(int i=0; i<n; i++) {
            for(int j=0; j<n; j++) {
                if(graph[i][u] + graph[v][j] + c < graph[i][j]) {
                    graph[i][j] = graph[i][u] + graph[v][j] + c;
                }
            }
        }
    }
    
    public int shortestPath(int node1, int node2) {
        return graph[node1][node2] == (int)1e8 ? -1 : graph[node1][node2];
    }
}

/**
 * Your Graph object will be instantiated and called as such:
 * Graph obj = new Graph(n, edges);
 * obj.addEdge(edge);
 * int param_2 = obj.shortestPath(node1,node2);
 */

File: 2678-design-graph-with-shortest-path-calculator/design-graph-with-shortest-path-calculator.js
/**
 * @param {number} n
 * @param {number[][]} edges
 */
var Graph = function(n, edges) {
    this.graph = new Array(n);

    for(let i=0; i<n; i++) this.graph[i] = [];

    // let graph = this.graph;
    
    for(let [u, v, c] of edges) {
        // console.log(this.graph);
        this.graph[u].push([v, c]);
    }

};

/** 
 * @param {number[]} edge
 * @return {void}
 */
Graph.prototype.addEdge = function(edge) {
    // let graph = this.graph;

    this.graph[edge[0]].push([edge[1], edge[2]]);
};

/** 
 * @param {number} node1 
 * @param {number} node2
 * @return {number}
 */
Graph.prototype.shortestPath = function(node1, node2) {
    const heap = new MinPriorityQueue();
    heap.enqueue({node: node1, cost: 0}, 0);
    let visited = new Set();    

    while(heap.size() > 0) {
        let top = heap.dequeue().element;
        if(visited.has(top.node)) {
            continue;
        }
        visited.add(top.node);
        if(top.node === node2) {
            return top.cost;
        }
        let next = this.graph[top.node];
        for(let neigh of next) {
            heap.enqueue({node: neigh[0], cost: top.cost + neigh[1]}, top.cost + neigh[1]);
        }
    }
    return -1;
};

/** 
 * Your Graph object will be instantiated and called as such:
 * var obj = new Graph(n, edges)
 * obj.addEdge(edge)
 * var param_2 = obj.shortestPath(node1,node2)
 */